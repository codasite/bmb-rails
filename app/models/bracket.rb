class Bracket < ApplicationRecord
  validates :title, presence: true
  validates :slug, presence: true, uniqueness: true
  validates :num_teams, presence: true, numericality: { greater_than: 0 }

  before_validation :generate_slug, if: :title_changed?

  serialize :round_names, coder: JSON

  scope :by_year, ->(year) { where(year: year) if year.present? }
  scope :by_month, ->(month) { where(month: month) if month.present? }
  scope :voting, -> { where(is_voting: true) }
  scope :tournament, -> { where(is_voting: false) }

  has_many :bracket_matches, dependent: :destroy
  has_many :picks, dependent: :destroy
  has_many :plays, dependent: :destroy

  def to_param
    slug
  end

  def voting?
    is_voting
  end

  def tournament?
    !is_voting
  end

  def has_fee?
    fee.present? && fee > 0
  end

  def current_round
    live_round_index
  end

  def completed?
    # Logic to determine if bracket is completed
    # This would depend on your specific business rules
    false
  end

  private

  def generate_slug
    self.slug = title.parameterize if title.present?
  end
end
