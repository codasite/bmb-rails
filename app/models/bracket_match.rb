class BracketMatch < ApplicationRecord
  belongs_to :bracket
  belongs_to :team1, class_name: 'Team', optional: true
  belongs_to :team2, class_name: 'Team', optional: true
  belongs_to :winner, class_name: 'Team', optional: true

  has_many :picks, dependent: :destroy

  validates :round, presence: true
  validates :position, presence: true

  scope :by_round, ->(round) { where(round: round) }
  scope :completed, -> { where.not(winner: nil) }
  scope :pending, -> { where(winner: nil) }

  def completed?
    winner.present?
  end

  def pending?
    !completed?
  end

  def teams
    [team1, team2].compact
  end

  def has_both_teams?
    team1.present? && team2.present?
  end
end
