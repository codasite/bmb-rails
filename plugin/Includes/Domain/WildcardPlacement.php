<?php

namespace WStrategies\BMB\Includes\Domain;

/**
 * Constants for wildcard placement options in brackets.
 */
class WildcardPlacement {
  const TOP = 0;
  const BOTTOM = 1;
  const CENTER = 2;
  const SPLIT = 3;

  const OPTIONS = [
    'top' => self::TOP,
    'bottom' => self::BOTTOM,
    'center' => self::CENTER,
    'split' => self::SPLIT,
  ];
}
