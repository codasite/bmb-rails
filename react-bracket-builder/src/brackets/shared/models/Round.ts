import { Nullable } from '../../../utils/types'
import { MatchNode } from './operations/MatchNode'
import { MatchRepr } from '../api/types/bracket'
import { Team } from './Team'

export class Round {
  index: number
  depth: number
  matches: Array<Nullable<MatchNode>>
  constructor(
    index: number,
    depth: number,
    matches: Array<Nullable<MatchNode>> = []
  ) {
    this.index = index
    this.depth = depth
    this.matches = matches
  }
  allPicked(): boolean {
    return this.matches.every((match) => {
      if (match === null) {
        return true
      }
      return match.team1Wins || match.team2Wins
    })
  }
  getTeam(teamId: number): Nullable<Team> {
    for (const [index, match] of this.matches.entries()) {
      if (match === null) {
        continue
      }
      const team1 = match.getTeam1()
      const team2 = match.getTeam2()

      if (team1?.id === teamId) {
        team1.side = index < this.matches.length / 2 ? 'left' : 'right'
        return team1
      }

      if (team2?.id === teamId) {
        team2.side = index < this.matches.length / 2 ? 'left' : 'right'
        return team2
      }
    }

    return null
  }
  serialize(): Nullable<MatchRepr>[] {
    return this.matches.map((match, i) => {
      if (match === null) {
        return null
      }
      return match.serialize()
    })
  }
}
