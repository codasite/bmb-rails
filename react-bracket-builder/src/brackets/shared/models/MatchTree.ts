import { Nullable } from '../../../utils/types'
import { BracketRes, MatchPick, MatchReq, MatchTreeRepr } from '../api'
import { WildcardPlacement } from './WildcardPlacement'
import { Team } from './Team'
import { MatchNode } from './operations/MatchNode'
import { Round } from './Round'
import { matchReprFromNumTeams } from './operations/MatchReprFromNumTeams'
import { getNumRounds } from './operations/GetNumRounds'
import { matchReprFromRes } from './operations/MatchReprFromRes'
import { linkNodes } from './operations/LinkNodes'

export class MatchTree {
  rounds: Round[]
  private numTeams: number
  private wildcardPlacement?: WildcardPlacement
  isVoting: boolean
  liveRoundIndex: number

  constructor(
    rounds: Round[] = [],
    wildcardPlacement?: WildcardPlacement,
    isVoting: boolean = false,
    liveRoundIndex: number = 0
  ) {
    linkNodes(rounds)
    this.rounds = rounds
    this.wildcardPlacement = wildcardPlacement
    this.isVoting = isVoting
    this.liveRoundIndex = liveRoundIndex
  }

  getRootMatch(): Nullable<MatchNode> {
    const lastRound = this.rounds[this.rounds.length - 1]
    if (!lastRound) {
      return null
    }
    return lastRound.matches[0]
  }

  clone(): MatchTree {
    return MatchTree.deserialize(this.serialize())
  }

  equals(other: MatchTree): boolean {
    return (
      JSON.stringify(this.serialize()) === JSON.stringify(other.serialize())
    )
  }

  /**
   * Returns the total number of POSSIBLE teams in the tournament.
   * This should reflect the number of teams that would be used to create the same structure with MatchTree.fromNumTeams()
   */
  getNumTeams(): number {
    if (this.numTeams) {
      return this.numTeams
    }
    // count all non null matches in the tree
    const numMatches = this.rounds.reduce((acc, round) => {
      return (
        acc +
        round.matches.reduce((acc, match) => {
          if (match) {
            return acc + 1
          }
          return acc
        }, 0)
      )
    }, 0)

    const numTeams = numMatches + 1
    this.numTeams = numTeams

    return numTeams
  }

  getWildcardPlacement(): WildcardPlacement | undefined {
    return this.wildcardPlacement
  }

  serialize(): MatchTreeRepr {
    const tree = this
    const rounds = tree.rounds.map((round) => {
      return round.serialize()
    })
    return {
      rounds: rounds,
      wildcardPlacement: tree.wildcardPlacement,
      isVoting: tree.isVoting,
      liveRoundIndex: tree.liveRoundIndex,
    }
  }

  advanceTeam = (
    roundIndex: number,
    matchIndex: number,
    left: boolean,
    requireComplete: boolean = false
  ) => {
    if (requireComplete && roundIndex > 0) {
      const prevRound = this.rounds[roundIndex - 1]
      if (prevRound && !prevRound.allPicked()) {
        return
      }
    }
    const round = this.rounds[roundIndex]
    const match = round.matches[matchIndex]
    if (!match) {
      return
    }
    match.team1Wins = false
    match.team2Wins = false
    if (left) {
      match.team1Wins = true
    } else {
      match.team2Wins = true
    }
  }

  anyPicked = (): boolean => {
    return this.rounds.some((round) => {
      return round.matches.some((match) => {
        if (!match) {
          return false
        }
        return match.team1Wins || match.team2Wins
      })
    })
  }

  allPicked = (): boolean => {
    if (this.isVoting) {
      return this.rounds[this.liveRoundIndex].allPicked()
    }
    return this.rounds.every((round) => {
      return round.allPicked()
    })
  }

  allTeamsAdded = (): boolean => {
    return this.everyMatch((match) => {
      let hasNeededTeams = true
      if (!match.left && !match.getTeam1()?.name) {
        hasNeededTeams = false
      }
      if (hasNeededTeams && !match.right && !match.getTeam2()?.name) {
        hasNeededTeams = false
      }
      return hasNeededTeams
    })
  }

  toMatchReq = (): MatchReq[] => {
    const matches: MatchReq[] = []
    this.forEachMatch((match, matchIndex, roundIndex) => {
      const matchReq: MatchReq = {
        roundIndex: match.roundIndex,
        matchIndex: match.matchIndex,
      }
      if (!match.left) {
        matchReq.team1 = match.getTeam1()?.toTeamReq()
      }
      if (!match.right) {
        matchReq.team2 = match.getTeam2()?.toTeamReq()
      }
      if (matchReq.team1 || matchReq.team2) {
        matches.push(matchReq)
      }
    })
    return matches
  }

  toMatchPicks = (): MatchPick[] => {
    const picks: MatchPick[] = []
    this.rounds.forEach((round, roundIndex) => {
      round.matches.forEach((match, matchIndex) => {
        if (!match) {
          return
        }
        const { team1Wins, team2Wins } = match
        const team1 = match.getTeam1()
        const team2 = match.getTeam2()
        const team1Winner = team1Wins && team1
        const team2Winner = team2Wins && team2

        if (team1Winner && team1.id) {
          picks.push({
            roundIndex,
            matchIndex,
            winningTeamId: team1.id,
          })
        } else if (team2Winner && team2.id) {
          picks.push({
            roundIndex,
            matchIndex,
            winningTeamId: team2.id,
          })
        }
      })
    })
    return picks
  }

  forEachMatch = (
    callback: (match: MatchNode, matchIndex: number, roundIndex: number) => void
  ) => {
    this.rounds.forEach((round, roundIndex) => {
      round.matches.forEach((match, matchIndex) => {
        if (!match) {
          return
        }
        callback(match, matchIndex, roundIndex)
      })
    })
  }

  everyMatch = (
    callback: (
      match: MatchNode,
      matchIndex: number,
      roundIndex: number
    ) => boolean
  ) => {
    return this.rounds.every((round, roundIndex) => {
      return round.matches.every((match, matchIndex) => {
        if (!match) {
          return true
        }
        return callback(match, matchIndex, roundIndex)
      })
    })
  }

  findMatch = (
    callback: (match: Nullable<MatchNode>) => boolean
  ): Nullable<MatchNode> => {
    let foundMatch: Nullable<MatchNode> = null
    this.rounds.some((round) => {
      return round.matches.some((match) => {
        if (callback(match)) {
          foundMatch = match
          return true
        }
        return false
      })
    })
    return foundMatch
  }

  static fromNumTeams(
    numTeams: number,
    wildcardPlacement: WildcardPlacement = WildcardPlacement.Top
  ): MatchTree {
    const matches = matchReprFromNumTeams(numTeams, wildcardPlacement)
    return MatchTree.deserialize({
      rounds: matches,
      wildcardPlacement,
    })
  }

  static fromMatchRes(
    bracket: Pick<
      BracketRes,
      'matches' | 'numTeams' | 'isVoting' | 'liveRoundIndex'
    >,
    wildcardPlacement?: WildcardPlacement
  ): MatchTree | null {
    const numRounds = getNumRounds(bracket.numTeams)

    const nestedMatches = matchReprFromRes(numRounds, bracket.matches)
    return MatchTree.deserialize({
      rounds: nestedMatches,
      wildcardPlacement,
      isVoting: bracket.isVoting,
      liveRoundIndex: bracket.liveRoundIndex,
    })
  }

  static fromPicks(
    bracket: Pick<
      BracketRes,
      'matches' | 'numTeams' | 'isVoting' | 'liveRoundIndex'
    >,
    picks: MatchPick[],
    wildcardPlacement?: WildcardPlacement
  ): MatchTree | null {
    const matchTree = MatchTree.fromMatchRes(bracket, wildcardPlacement)
    if (!matchTree) {
      return null
    }
    for (const pick of picks) {
      const { roundIndex, matchIndex, winningTeamId } = pick
      const match = matchTree.rounds[roundIndex].matches[matchIndex]
      if (!match) {
        throw new Error('Match not found')
      }
      match._pick = pick

      const team1 = match.getTeam1()
      const team2 = match.getTeam2()
      if (team1?.id === winningTeamId) {
        match.team1Wins = true
      } else if (team2?.id === winningTeamId) {
        match.team2Wins = true
      } else {
        throw new Error('Match does not have winningTeam from pick')
      }
    }
    return matchTree
  }

  static deserialize(matchTreeRepr: MatchTreeRepr): MatchTree {
    const { rounds: matchRes, wildcardPlacement } = matchTreeRepr
    const rounds = matchRes.map((round, roundIndex) => {
      const depth = matchRes.length - roundIndex - 1
      const newRound = new Round(roundIndex, depth)
      const matches = round.map((match, matchIndex) => {
        if (match === null) {
          return null
        }
        const {
          id,
          team1: team1Repr,
          team2: team2Repr,
          team1Wins,
          team2Wins,
          pick,
        } = match
        const team1 = team1Repr ? new Team(team1Repr.name, team1Repr.id) : null
        const team2 = team2Repr ? new Team(team2Repr.name, team2Repr.id) : null
        // const newMatch = new MatchNode(roundIndex, matchIndex, depth, match.id, team1, team2, team1Wins, team2Wins)
        const newMatch = new MatchNode({
          id,
          matchIndex,
          roundIndex,
          team1,
          team2,
          team1Wins,
          team2Wins,
          left: null,
          right: null,
          parent: null,
          depth,
          pick,
        })
        return newMatch
      })
      newRound.matches = matches
      return newRound
    })
    const tree = new MatchTree(
      rounds,
      wildcardPlacement,
      matchTreeRepr.isVoting,
      matchTreeRepr.liveRoundIndex
    )
    return tree
  }
}
