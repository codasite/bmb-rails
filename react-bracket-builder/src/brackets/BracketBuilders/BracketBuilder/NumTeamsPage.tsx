import React, { useEffect, useState } from 'react'
import { NumTeamsPicker } from './NumTeamsPicker'
import { MatchTree } from '../../shared/models/MatchTree'
import darkBracketBg from '../../shared/assets/bracket-bg-dark.png'
import { BracketPreview } from './BracketPreview'
import { isPowerOfTwo } from '../../shared/components/Bracket/utils'
import { WildcardPicker } from './WildcardPicker'
import { ActionButton } from '../../shared/components/ActionButtons'
import { BracketMeta } from '../../shared/context/context'
import { WildcardPlacement } from '../../shared/models/WildcardPlacement'
import { BracketTitle } from './BracketTitle'

export interface NumTeamsPickerState {
  currentValue: number
  selected: boolean
}
interface NumTeamsPageProps {
  matchTree?: MatchTree
  setMatchTree?: (matchTree: MatchTree) => void
  onAddTeamsClick: () => void
  numTeams: number
  setNumTeams: (numTeams: number) => void
  teamPickerDefaults: number[]
  teamPickerMin: number[]
  teamPickerMax: number[]
  wildcardPlacement: WildcardPlacement
  setWildcardPlacement: (placement: WildcardPlacement) => void
  teamPickerState: NumTeamsPickerState[]
  setTeamPickerState: (state: NumTeamsPickerState[]) => void
  bracketMeta: BracketMeta
  setBracketMeta?: (bracketMeta: BracketMeta) => void
}

export const NumTeamsPage = (props: NumTeamsPageProps) => {
  const {
    matchTree,
    setMatchTree,
    onAddTeamsClick,
    bracketMeta,
    setBracketMeta,
    numTeams,
    setNumTeams,
    teamPickerDefaults,
    teamPickerMin,
    teamPickerMax,
    wildcardPlacement,
    setWildcardPlacement,
    teamPickerState,
    setTeamPickerState,
  } = props

  const [showTitleError, setShowTitleError] = useState(false)

  // Update the global `numTeams` variable whenever picker state changes
  useEffect(() => {
    const picker = teamPickerState.find((picker) => picker.selected)
    if (picker) {
      const numTeams = picker.currentValue
      setNumTeams(numTeams)
      if (setMatchTree) {
        setMatchTree(MatchTree.fromNumTeams(numTeams, wildcardPlacement))
      }
    }
  }, [teamPickerState, wildcardPlacement])

  const updateTeamPicker = (index: number, newPicker: NumTeamsPickerState) => {
    const newPickers = teamPickerState.map((picker, i) => {
      if (i === index) {
        return newPicker
      }
      return picker
    })
    setTeamPickerState(newPickers)
  }

  const updateTeamPickerValueBy = (index: number, value: number) => {
    const picker = teamPickerState[index]
    const newPicker = {
      ...picker,
      currentValue: picker.currentValue + value,
    }
    updateTeamPicker(index, newPicker)
  }

  const incrementTeamPicker = (index: number) => {
    updateTeamPickerValueBy(index, 1)
  }

  const decrementTeamPicker = (index: number) => {
    updateTeamPickerValueBy(index, -1)
  }

  const setTeamPickerValue = (index: number, value: number) => {
    const picker = teamPickerState[index]
    const newPicker = {
      ...picker,
      currentValue: value,
    }
    updateTeamPicker(index, newPicker)
  }

  const setTeamPickerSelected = (
    index: number,
    resetValue: boolean = false
  ) => {
    const newPickers = teamPickerState.map((picker, i) => {
      if (i === index) {
        return {
          ...picker,
          currentValue: resetValue
            ? teamPickerDefaults[index]
            : picker.currentValue,
          selected: true,
        }
      }
      return {
        ...picker,
        selected: false,
      }
    })
    setTeamPickerState(newPickers)
  }

  /**
   * Get the function to select the next team picker
   * When the next team picker is selected, its value is reset to the default value
   */
  const getSelectNextTeamPicker = (index: number) => {
    if (index < teamPickerState.length - 1) {
      return () => setTeamPickerSelected(index + 1, true)
    }
    return undefined
  }

  /**
   * Get the function to select the previous team picker
   * When the previous team picker is selected, its value is reset to the default value
   */
  const getSelectPrevTeamPicker = (index: number) => {
    if (index > 0) {
      return () => setTeamPickerSelected(index - 1, true)
    }
    return undefined
  }

  // Show wild card options if numTeams is a power of 2
  const showWildCardOptions = !isPowerOfTwo(numTeams)

  const handleWildcardPlacement = (placement: WildcardPlacement) => {
    setWildcardPlacement(placement)
  }

  const handleAddTeamsClick = () => {
    if (!bracketMeta.title) {
      setShowTitleError(true)
      // scroll to top of page
      window.scrollTo(0, 0)
    } else {
      onAddTeamsClick()
    }
  }

  const onTitleChange = (event: any) => {
    setShowTitleError(false)
  }

  return (
    <div
      className="tw-bg-no-repeat tw-bg-top tw-bg-cover"
      style={{ backgroundImage: `url(${darkBracketBg})` }}
    >
      <div className="tw-flex tw-flex-col tw-gap-40 tw-pb-[240px] tw-pt-60 tw-max-w-screen-lg tw-m-auto tw-px-20 lg:tw-px-0">
        <BracketTitle
          title={bracketMeta.title}
          placeholder="Name Your Bracket"
          setTitle={(title) => setBracketMeta({ ...bracketMeta, title })}
          onChange={onTitleChange}
          showError={showTitleError}
          errorMessage="Bracket name is required."
        />
        {matchTree && (
          <div>
            <BracketPreview matchTree={matchTree} />
          </div>
        )}
        <div
          className={`tw-flex tw-flex-col tw-gap-24 ${
            showWildCardOptions ? '' : 'tw-pb-24'
          }`}
        >
          <span className="tw-text-white/50 tw-text-center tw-font-500 tw-text-16 md:tw-text-24">
            How Many Teams in Your Bracket?
          </span>
          <div className="tw-flex tw-flex-col md:tw-flex-row tw-gap-24">
            {teamPickerState.map((pickerState, i) => {
              return (
                <NumTeamsPicker
                  currentValue={pickerState.currentValue}
                  defaultValue={teamPickerDefaults[i]}
                  min={teamPickerMin[i]}
                  max={teamPickerMax[i]}
                  selected={pickerState.selected}
                  setSelected={() => setTeamPickerSelected(i)}
                  increment={() => incrementTeamPicker(i)}
                  decrement={() => decrementTeamPicker(i)}
                  setCurrentValue={(value) => setTeamPickerValue(i, value)}
                  selectNextPicker={getSelectNextTeamPicker(i)}
                  selectPrevPicker={getSelectPrevTeamPicker(i)}
                  key={i}
                />
              )
            })}
          </div>
          {showWildCardOptions && (
            <WildcardPicker
              wildcardPlacement={wildcardPlacement}
              onWildcardPlacementChanged={handleWildcardPlacement}
            />
          )}
        </div>
        {/* <button className='tw-rounded-8 tw-border tw-border-solid tw-border-green tw-bg-green/15 tw-p-16 tw-flex tw-justify-center tw-cursor-pointer' onClick={onAddTeamsClick}>
          <span className='tw-text-white tw-font-500 tw-text-20 tw-uppercase tw-font-sans '>Add Your Teams</span>
        </button> */}
        <ActionButton variant="green" onClick={handleAddTeamsClick}>
          <span className="tw-text-white tw-font-500 tw-text-20 tw-uppercase tw-font-sans">
            Add Your Teams
          </span>
        </ActionButton>
      </div>
    </div>
  )
}
