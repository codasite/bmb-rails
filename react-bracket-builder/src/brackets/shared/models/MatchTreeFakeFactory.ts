import { MatchTree } from './MatchTree'

export function fourTeamMatchTree() {
  const matches = [
    [
      {
        id: 9,
        roundIndex: 0,
        matchIndex: 0,
        team1: { id: 17, name: 'Team 1' },
        team2: { id: 18, name: 'Team 2' },
        team1Wins: true,
      },
      {
        id: 10,
        roundIndex: 0,
        matchIndex: 1,
        team1: { id: 19, name: 'Team 3' },
        team2: { id: 20, name: 'Team 4' },
        team1Wins: true,
      },
    ],
    [{ roundIndex: 1, matchIndex: 1, team2Wins: true }],
  ]
  return MatchTree.deserialize({ rounds: matches })
}
