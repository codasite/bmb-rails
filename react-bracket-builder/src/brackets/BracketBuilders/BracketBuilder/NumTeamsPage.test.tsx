import { render } from '@testing-library/react'
import { NumTeamsPage } from './NumTeamsPage'
import {
  teamPickerDefaults,
  teamPickerMax,
  teamPickerMin,
} from './BracketBuilder'
import { WildcardPlacement } from '../../shared/models/WildcardPlacement'
import { fourTeamMatchTree } from '../../shared/models/MatchTreeFakeFactory'

describe('NumTeamsPage', () => {
  test('renders page correctly', () => {
    const noop = () => {}
    const { asFragment } = render(
      <NumTeamsPage
        numTeams={16}
        setNumTeams={noop}
        onAddTeamsClick={noop}
        setTeamPickerState={noop}
        teamPickerDefaults={teamPickerDefaults}
        teamPickerMin={teamPickerMin}
        teamPickerMax={teamPickerMax}
        wildcardPlacement={WildcardPlacement.Center}
        teamPickerState={teamPickerDefaults.map((val, i) => ({
          currentValue: val,
          selected: i === 0,
        }))}
        setWildcardPlacement={noop}
        matchTree={fourTeamMatchTree()}
        bracketMeta={{ title: 'Test bracket', date: 'test date' }}
      />
    )
    expect(asFragment()).toMatchSnapshot()
  })
})
