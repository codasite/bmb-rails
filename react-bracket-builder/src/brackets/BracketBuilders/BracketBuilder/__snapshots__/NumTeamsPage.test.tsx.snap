// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NumTeamsPage renders page correctly 1`] = `
<DocumentFragment>
  <div
    class="tw-bg-no-repeat tw-bg-top tw-bg-cover"
    style="background-image: url(undefined);"
  >
    <div
      class="tw-flex tw-flex-col tw-max-w-screen-lg tw-m-auto tw-px-20 lg:tw-px-0 tw-pb-100 tw-pt-60 "
    >
      <div
        class="tw-py-30"
      >
        <div
          class="tw-flex tw-flex-col tw-items-center tw-gap-20"
        >
          <div
            class="tw-h-[50px] tw-w-[200px] md:tw-h-[75px] md:tw-w-[300px] lg:tw-h-[85px] lg:tw-w-[340px] tw-bg-[url('https://s3.amazonaws.com/backmybracket.com/bmb_text_logo_500.png')] tw-bg-no-repeat tw-bg-contain"
          />
          <h1
            class="tw-text-32 sm:tw-text-48 md:tw-text-64 lg:tw-text-80 tw-text-center"
          >
            Bracket Builder
          </h1>
        </div>
      </div>
      <div
        class="tw-flex tw-flex-col tw-gap-40 "
      >
        <div
          class="tw-flex tw-justify-center tw-border-b-solid tw-border-white/30 tw-p-16 "
        >
          <h1
            class="tw-font-500 tw-text-24 sm:tw-text-32 !tw-text-white/20 tw-text-center"
          >
            Test bracket
          </h1>
        </div>
        <div>
          <div
            class="tw-flex tw-flex-col tw-justify-center tw-m-auto tw-max-w-[474px] tw-h-[296px]"
          >
            <div
              class="tw-flex tw-justify-between tw-gap-8 md:tw-gap-16 "
            >
              <div
                class="tw-flex tw-flex-col tw-justify-center tw-flex-grow"
              >
                <div
                  class="tw-flex tw-flex-col tw-gap-[8px]"
                >
                  <div
                    class="tw-rounded-4 tw-h-[24px] tw-max-w-[150px] tw-bg-white/30"
                  />
                  <div
                    class="tw-rounded-4 tw-h-[24px] tw-max-w-[150px] tw-bg-white/30"
                  />
                </div>
              </div>
              <div
                class="tw-flex tw-flex-col tw-justify-center tw-flex-grow"
              >
                <div
                  class="tw-flex tw-flex-col tw-gap-[8px] tw-pb-16"
                >
                  <div
                    class="tw-rounded-4 tw-h-[24px] tw-max-w-[150px] tw-bg-white/10"
                  />
                  <div
                    class="tw-rounded-4 tw-h-[24px] tw-max-w-[150px] tw-bg-white/10"
                  />
                </div>
              </div>
              <div
                class="tw-flex tw-flex-col tw-justify-center tw-flex-grow"
              >
                <div
                  class="tw-flex tw-flex-col tw-gap-[8px]"
                >
                  <div
                    class="tw-rounded-4 tw-h-[24px] tw-max-w-[150px] tw-bg-white/30"
                  />
                  <div
                    class="tw-rounded-4 tw-h-[24px] tw-max-w-[150px] tw-bg-white/30"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="tw-flex tw-flex-col tw-gap-24 tw-pb-24"
        >
          <span
            class="tw-text-white/50 tw-text-center tw-font-500 tw-text-16 md:tw-text-24"
          >
            How Many Teams in Your Bracket?
          </span>
          <div
            class="tw-flex tw-flex-col md:tw-flex-row tw-gap-24"
          >
            <div
              class="tw-flex tw-flex-col tw-gap-24 tw-grow"
            >
              <div
                class="tw-flex tw-justify-center tw-h-[136px] tw-items-center tw-border-solid tw-rounded-8 tw-relative tw-cursor-pointer tw-bg-green/15 tw-border-4 tw-border-green"
              >
                <span
                  class="tw-font-500 tw-text-48 tw-text-white"
                >
                  16
                </span>
                <span
                  class="tw-absolute tw-bottom-8 tw-left-14 tw-text-green tw-font-500 tw-text-12"
                >
                  Default
                </span>
              </div>
              <div
                class="tw-flex tw-justify-center tw-gap-12"
              >
                <button
                  class="tw-flex tw-justify-center tw-items-center tw-flex-grow tw-border-solid tw-border tw-rounded-4 tw-text-36 tw-font-500 tw-font-sans tw-h-30 tw-bg-transparent tw-border-white tw-text-white tw-cursor-pointer"
                >
                  -
                </button>
                <button
                  class="tw-flex tw-justify-center tw-items-center tw-flex-grow tw-border-solid tw-border tw-rounded-4 tw-text-36 tw-font-500 tw-font-sans tw-h-30 tw-bg-transparent tw-border-white tw-text-white tw-cursor-pointer"
                >
                  +
                </button>
              </div>
            </div>
            <div
              class="tw-flex tw-flex-col tw-gap-24 tw-grow"
            >
              <div
                class="tw-flex tw-justify-center tw-h-[136px] tw-items-center tw-border-solid tw-rounded-8 tw-relative tw-cursor-pointer tw-border tw-border-white/50"
              >
                <span
                  class="tw-font-500 tw-text-48 tw-text-white"
                >
                  32
                </span>
              </div>
            </div>
            <div
              class="tw-flex tw-flex-col tw-gap-24 tw-grow"
            >
              <div
                class="tw-flex tw-justify-center tw-h-[136px] tw-items-center tw-border-solid tw-rounded-8 tw-relative tw-cursor-pointer tw-border tw-border-white/50"
              >
                <span
                  class="tw-font-500 tw-text-48 tw-text-white"
                >
                  64
                </span>
              </div>
            </div>
          </div>
        </div>
        <button
          class="tw-flex tw-flex-row tw-items-center tw-justify-center tw-font-sans tw-uppercase tw-whitespace-nowrap tw-cursor-pointer disabled:tw-cursor-default tw-leading-none tw-cursor-pointer tw-bg-green/15 tw-text-white tw-gap-10 tw-text-20 tw-font-500 tw-rounded-8 tw-py-16 tw-border-solid tw-border tw-border-green"
        >
          <span
            class="tw-text-white tw-font-500 tw-text-20 tw-uppercase tw-font-sans"
          >
            Add Your Teams
          </span>
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
